kaiax/gov/param.go

```diff
@@ -0,0 +1,442 @@
+		Canonicalizer:     boolCanonicalizer,
+		FormatChecker:     noopFormatChecker,
+		DefaultValue:      false,
+		VoteForbidden:     true,
+	},
+	RewardKip82Ratio: {
+		Name:              "reward.kip82ratio",
+		ParamSetFieldName: "Kip82Ratio",
+		Canonicalizer:     stringCanonicalizer,
+		FormatChecker: func(cv interface{}) bool {
+			v, ok := cv.(string)
+			if !ok {
+				return false
+			}
+			parts := strings.Split(v, "/")
+			if len(parts) != 2 {
+				return false
+			}
+			sum := 0
+			for _, part := range parts {
+				num, err := strconv.Atoi(part)
+				if err != nil {
+					return false
+				}
+				if num < 0 {
+					return false
+				}
+				sum += num
+			}
+
+			return sum == 100
+		},
+		DefaultValue:  "20/80",
+		VoteForbidden: false,
+	},
+	RewardMintingAmount: {
+		Name:              "reward.mintingamount",
+		ParamSetFieldName: "MintingAmount",
+		Canonicalizer:     bigIntCanonicalizer,
+		FormatChecker:     noopFormatChecker,
+		DefaultValue:      big.NewInt(0),
+		VoteForbidden:     false,
+	},
+	RewardMinimumStake: {
+		Name:              "reward.minimumstake",
+		ParamSetFieldName: "MinimumStake",
+		Canonicalizer:     bigIntCanonicalizer,
+		FormatChecker: func(cv interface{}) bool {
+			v, ok := cv.(*big.Int)
+			if !ok {
+				return false
+			}
+			return v.Sign() >= 0
+		},
+		DefaultValue:  big.NewInt(2000000),
+		VoteForbidden: true,
+	},
+	RewardProposerUpdateInterval: {
+		Name:              "reward.proposerupdateinterval",
+		ParamSetFieldName: "ProposerUpdateInterval",
+		Canonicalizer:     uint64Canonicalizer,
+		FormatChecker:     noopFormatChecker,
+		DefaultValue:      uint64(3600),
+		VoteForbidden:     true,
+	},
+	RewardRatio: {
+		Name:              "reward.ratio",
+		ParamSetFieldName: "Ratio",
+		Canonicalizer:     stringCanonicalizer,
+		FormatChecker: func(cv interface{}) bool {
+			v, ok := cv.(string)
+			if !ok {
+				return false
+			}
+			parts := strings.Split(v, "/")
+			if len(parts) != 3 {
+				return false
+			}
+			sum := 0
+			for _, part := range parts {
+				num, err := strconv.Atoi(part)
+				if err != nil {
+					return false
+				}
+				if num < 0 {
+					return false
+				}
+				sum += num
+			}
+
+			return sum == 100
+		},
+		DefaultValue:  "100/0/0",
+		VoteForbidden: false,
+	},
+	RewardStakingUpdateInterval: {
+		Name:              "reward.stakingupdateinterval",
+		ParamSetFieldName: "StakingUpdateInterval",
+		Canonicalizer:     uint64Canonicalizer,
+		FormatChecker:     noopFormatChecker,
+		DefaultValue:      uint64(86400),
+		VoteForbidden:     true,
+	},
+	RewardUseGiniCoeff: {
+		Name:              "reward.useginicoeff",
+		ParamSetFieldName: "UseGiniCoeff",
+		Canonicalizer:     boolCanonicalizer,
+		FormatChecker:     noopFormatChecker,
+		DefaultValue:      false,
+		VoteForbidden:     true,
+	},
+}
+
+var ParamNameToEnum map[string]ParamEnum
+
+func init() {
+	ParamNameToEnum = make(map[string]ParamEnum)
+	for k, v := range Params {
+		ParamNameToEnum[v.Name] = k
+	}
+}
```

Comment:
I want to avoid the situation where I added a param but forgot to update `ParamNameToEnum` as well.
If I can ensure this somehow, I'm fine with flattening it.

Reply:
Also, I should tell you I'm going to use a string-enum (I'm not sure what to call this), and remove the need for integer types altogether.

```
const (
  GovModeName ParamName = "governance.governancemode"
  GovNodeName           = "governance.governingnode"
)
```

Reply:
Got it, thanks.